v1/protocol_binding/pubsub_binder_test.cc

```diff
@@ -0,0 +1,268 @@
+  ASSERT_EQ((*bind).attributes().at("ce-source"), "2");
+  ASSERT_EQ((*bind).attributes().at("ce-specversion"), "3");
+  ASSERT_EQ((*bind).attributes().at("ce-type"), "4");
+  ASSERT_EQ((*bind).data(), "aGVsbG8=");
+}
+
+TEST_F(BindTest, Structured_Required) {
+  StatusOr<PubsubMessage> bind = binder.Bind(ce, Format::kJson);
+
+  ASSERT_TRUE(bind.ok());
+  ASSERT_EQ((*bind).attributes().at("content-type"), "application/cloudevents+json");
+  ASSERT_EQ((*bind).data(), "{\n\t\"id\" : \"1\",\n\t\"source\" : \"2\",\n\t\"specversion\" : \"3\",\n\t\"type\" : \"4\"\n}");
+}
+
+TEST_F(BindTest, Structured_Optional) {
+  CloudEvent_CloudEventAttribute opt_val;
+  opt_val.set_ce_string("5");
+  (*ce.mutable_attributes())["opt"] = opt_val;
+
+  StatusOr<PubsubMessage> bind = binder.Bind(ce, Format::kJson);
+
+  ASSERT_TRUE(bind.ok());
+  ASSERT_EQ((*bind).attributes().at("content-type"), "application/cloudevents+json");
+  ASSERT_EQ((*bind).data(), "{\n\t\"id\" : \"1\",\n\t\"opt\" : \"5\",\n\t\"source\" : \"2\",\n\t\"specversion\" : \"3\",\n\t\"type\" : \"4\"\n}");
+}
+
+TEST_F(BindTest, Structured_BinData) {
+  ce.set_binary_data("1010");
+
+  StatusOr<PubsubMessage> bind = binder.Bind(ce, Format::kJson);
+
+  ASSERT_TRUE(bind.ok());
+  ASSERT_EQ((*bind).attributes().at("content-type"), "application/cloudevents+json");
+  ASSERT_EQ((*bind).data(), "{\n\t\"data_base64\" : \"1010\",\n\t\"id\" : \"1\",\n\t\"source\" : \"2\",\n\t\"specversion\" : \"3\",\n\t\"type\" : \"4\"\n}");
+}
+
+TEST_F(BindTest, Structured_TextData) {
+  ce.set_text_data("hello");
+
+  StatusOr<PubsubMessage> bind = binder.Bind(ce, Format::kJson);
+
+  ASSERT_TRUE(bind.ok());
+  ASSERT_EQ((*bind).attributes().at("content-type"), "application/cloudevents+json");
+  ASSERT_EQ((*bind).data(), "{\n\t\"data\" : \"hello\",\n\t\"id\" : \"1\",\n\t\"source\" : \"2\",\n\t\"specversion\" : \"3\",\n\t\"type\" : \"4\"\n}");
+}
+
+TEST_F(UnbindBinaryTest, Invalid) {
+  PubsubBinder binder;
+  PubsubMessage pubsub_msg;
+  StatusOr<CloudEvent> unbind = binder.Unbind(pubsub_msg);
+
+  ASSERT_FALSE(unbind.ok());
+  ASSERT_TRUE(absl::IsInvalidArgument(unbind.status()));
+}
+
+TEST_F(UnbindBinaryTest, Required) {
+  StatusOr<CloudEvent> unbind = binder.Unbind(pubsub_msg);
+
+  ASSERT_TRUE(unbind.ok());
+  ASSERT_EQ((*unbind).id(), "1");
+  ASSERT_EQ((*unbind).source(), "2");
+  ASSERT_EQ((*unbind).spec_version(), "3");
+  ASSERT_EQ((*unbind).type(), "4");
+}
+
+TEST_F(UnbindBinaryTest, Optional) {
+  (*pubsub_msg.mutable_attributes())["ce-opt"] = "5";
+
+  StatusOr<CloudEvent> unbind = binder.Unbind(pubsub_msg);
+
+  ASSERT_TRUE(unbind.ok());
+  ASSERT_EQ((*unbind).id(), "1");
+  ASSERT_EQ((*unbind).source(), "2");
+  ASSERT_EQ((*unbind).spec_version(), "3");
+  ASSERT_EQ((*unbind).type(), "4");
+  ASSERT_EQ((*unbind).attributes().at("opt").ce_string(), "5");
+}
+
+TEST_F(UnbindBinaryTest, Data) {
+  (*pubsub_msg.mutable_attributes())["ce-opt"] = "5";
+  pubsub_msg.set_data("1010");
```

Comment:
This codepath is a little questionable to me. When populating the the pubsub message, we should put in the base64 encoded string and when unbound, the output should be the original bytes (not base-64 encoded).

Reply:
Ah okay, I just interpreted binary_data as base64 encoded data because of the definition of binary in the CE type system: 
https://github.com/cloudevents/spec/blob/master/spec.md#type-system, so i thought i could unbind the data to cloud_event.binary_data() with no decoding.

So instead I will decode the data to cloud_event.text_data()?

Reply:
Implemented change is in [this commit](https://github.com/googleinterns/cloudevents-sdk-cpp/pull/55/commits/24a4be1f2d079fe0dd1dda0f3fabfb78053063db)

Reply:
Based on our discussions I removed all encoding/ decoding for handling the PubsubMessage proto :) 

https://github.com/googleinterns/cloudevents-sdk-cpp/pull/55/commits/1238ef0d388ba727c419bbca8cffaf5981c4c0a6